### ISSUE TITLE
Memory corruption in IPC subsystem

### ISSUE THREAD

**Issue by mkow** [oscarlab/graphene#1724]
## Description of the problem

Graphene's IPC subsystem doesn't care too much about ownership of objects, which are freely passed to functions and saved in various structures. This sometimes leads to memory corruptions - one of which is triggered by our R example.

## Steps to reproduce

Check out https://github.com/mkow/graphene/tree/mkow/r-debugging (this branch has additional debugging info printed), then:

```Bash
cd Examples/r
make SGX=1 DEBUG=1
SGX=1 ./pal_loader R.manifest --slave --vanilla -f scripts/sample.r
```

## Expected results

All handles appearing in the log are valid.

## Actual results

```
[33:sh] Adding port (handle 0x83bafd52a:pal_type_pipesrv) for process 947 (type 0002)
[33:sh] Adding port (handle 0x83bad9df0:pal_type_process) for process 941 (type 0011)
...
[33:sh] __free_ipc_info:122 DkObjectClose(0x83bafd52a:pal_type_pipesrv)
[33:sh] __free_ipc_info:122 DkObjectClose(0x83bad9df0:pal_type_process)
...
[P947:sh] Listening to process 941 on port 0x83aaef0a0 (handle 0x83bad9df0:<unknown>, type 0011)
[P947:sh] Listening to process 947 on port 0x83aaef048 (handle 0x83bafd52a:<unknown>, type 0002)
```

0x83bad9df0 and 0x83bafd52a are dangling pointers, as one can see from the log.

## Bug analysis

This particular bug is probably caused by total lack of object ownership notion around `shim_ipc_info` and `shim_ipc_port` structures, specifically, `add_ipc_port_by_id()` seems to take ownership over its `hdl` argument, but this isn't documented anywhere and different callsites behave differently when using it. Most likely this PAL handle ends up also in a different structure, which gets freed while a corresponding `shim_ipc_port` still lives and is used.

**Comment by dimakuv** [comment]
I will try to fix this when I have time. In particular, we must merge `shim_ipc_port` and `shim_ipc_info` into one struct (removing the latter).

**Comment by dimakuv** [comment]
Submitted a fix #1799. However, it is far from what we expected to do/modify.

> In particular, we must merge shim_ipc_port and shim_ipc_info into one struct (removing the latter).

This may be a good refactoring PR, but it's not about this bug and it's not that simple. We general, we currently have 3 `shim_ipc_info` objects: for myself (sending IPC messages to myself and using this as a parent object for children), for my parent process, and for my leader process.

IPC ports are a bit different and more versatile: there can be several open ports to the same `shim_ipc_info` object (i.e., to the same process). So it's not a 1:1 mapping.

Anyway, the real reason for this particular bug is the fact that on `execve()`, we need to migrate the PAL handles (and this is done through the checkpointing of the new "child" `shim_process` object). But there is no refcounting of PAL handles in LibOS, and it seemed an overkill to add refcounting just for this particular use case.
